### ISSUE TITLE
fuzzy search history, bookmarks and open tabs

### ISSUE THREAD

**Issue by mohitleo9** [1995eaton/chromium-vim#91]

**Comment by JindrichPilar** [comment]
I think he meant what [CtrlP](https://github.com/kien/ctrlp.vim)  plugin does when searching. (Also Sublime Text and PhpStorm)
There is short [animation](http://dab1nmslvvntp.cloudfront.net/wp-content/uploads/2014/02/demo07.gif)  how it looks in extreme case. You write ``DC`` and it Matches ``D.*C.*``

Instead of exact match of inserted word it "tries to figure out good enough match". It is meant as shortcut when you know coupple of keywords and their order. It is much faster than writing regular expression and o can gets to it unbelievably quckly.

example:
Lets take URL to this issue: ``https://github.com/1995eaton/chromium-vim/issues/91``

I know it is issue on github and it is plugin for chrome, but don't remember exact name so I write:
``github/chrom/issue``
And hope autocomplete will suggest ``https://github.com/1995eaton/chromium-vim/issues/91``
Fuzzy find can be created smart, but simple implementation could simply transform it to:
```BASH
.*github.*/.*chrom.*/.*issue.*
```

Making it smarter is the harder part. For example you could do more searches each more general.
1. Exact match (exact string is there)
2. Match of the regexp mentioned above
3. Match with different order
4. Match without some of the words


I might make PR, but I don't have much time this month.

**Comment by 1995eaton** [comment]
Regex matching is enabled by default in certain modes (`:history` and `:bookmarks`). Is this the behavior you're looking for but in the `:open`/`:tabnew` commands?

**Comment by JindrichPilar** [comment]
No, fuzzy matching isn't regex matching. Fuzzy as a word coud be explained as ``not exact or incomplete``. Regex on the other hand is **very** exact. 
Fuzzy search takes input and tries to interpret it. 

You can use regex to implement it, but there would be component that would turn input into multiple separate regexular expressions automatically.

example:

User input |  Expected autocompletion
------------- | -------------
github/chrom/issue  | ``https://github.com/1995eaton/chromium-vim/issues/91``
FacMarkZuck  | ``https://www.facebook.com/zuck``
reddit/linux | ``https://www.reddit.com/r/linux/``
linux reddit | ``https://www.reddit.com/r/linux/``

**Comment by perpen** [comment]
Indeed I really miss this feature.
I would only add to Jindrich's description that it would be better for the completion to use both the URL and the titles of pages.

**Comment by waynehoover** [comment]
This would be an amazing feature to have.

**Comment by hiberabyss** [comment]
It seems that `.*gtest` can't match `GoogleTest(gtest)` in `:history`, I'm not sure whether I use regexp match in wrong way.

PS: Can you add new separator `[ ( "` like `<space>` in url match ?

**Comment by dumblob** [comment]
Any progress on this very handy feature?

**Comment by kurtontheway** [comment]
This would be a great feature to improve ux.
